core/src/main/java/contacts/core/Where.kt

```diff
@@ -403,108 +468,256 @@ private class RawContactsTableWhere(whereString: String) : Where<RawContactsFiel
+        options: String? = null
+    ) : this(
+        lhs = lhs as LeftHandSide,
+        operator = operator,
+        rhs = rhs as RightHandSide,
+        options = options
+    )
+
+    /**
+     * Construct a copy of this where with the new field type determined by [substituteField].
+     */
+    internal fun <R : Field> copyWithNewFieldType(substituteField: (Field) -> Field): Where<R> {
+        /*
+         * Okay. Time for some "recursion" hehehe =). You know, I can't believe this interview
+         * skill is actually coming in handy... for once LOL! Ohh I'm so excited to have encountered
+         * this problem in the REAL LIFE! Ohh, I'm so hyped! Anyways, this is probably an easy level
+         * question in leet code. Standard tree traversal. So, I'll translate this to a leet code
+         * question. It's essentially "find leaf nodes of a binary tree".
+         *
+         * Given the root node (this) of a binary tree, use the substituteField function to replace
+         * the leaf nodes. See the class documentation for the binary tree structure.
+         *
+         * Without further ado, here is the code!
+         */
+        return if (lhs is FieldHolder && rhs is ValueHolder) {
+            // Base case. Perform the substitution.
+            Where(
+                lhs = FieldHolder(substituteField(lhs.field)),
+                operator = operator,
+                rhs = rhs,
+                options = options
+            )
+        } else if (lhs is WhereHolder && rhs is WhereHolder) {
+            // Recursive case. Traverse tree.
+            Where(
+                lhs = WhereHolder(lhs.where.copyWithNewFieldType(substituteField)),
+                operator = operator,
+                rhs = WhereHolder(rhs.where.copyWithNewFieldType(substituteField)),
+                options = options
+            )
+        } else {
+            // This should not happen with the current structure. If this does happen, it means
+            // that we made some changes that broke the structure.
+            throw ContactsException(
+                "Unhandled Where form lhs: ${lhs.javaClass.simpleName}, rhs:${rhs.javaClass.simpleName}"
+            )
+        }
     }
 
-    if (field is DataField && field.mimeType.value.isNotBlank()) {
-        where += " AND ${Fields.MimeType.columnName} = '${field.mimeType.value}'"
+    // Only evaluate this once to save some CPU. This assumes that property values are immutable.
+    // If there are mutable property values, then this will be evaluated at the time of invocation
+    // and will not mutate along with the mutable property values (e.g. a mutable list). I don't
+    // think consumers expect this to mutate anyways if they happen to save a reference to it.
+    private val evaluatedWhereString: String by unsafeLazy {
+
+        var whereString = when (operator) {
+            is Operator.Combine -> "($lhs) $operator ($rhs)"
```

Comment:
With the current structure, the `Combine` operators (AND, OR) imply that the `lhs` and `rhs` are both un-evaluated `Where`s. So, this is another for of recursion at play 😁